ghmirror/core/mirror_requests.py

```diff
@@ -58,6 +58,64 @@ def _cache_response(resp, cache, cache_key):
         cache[cache_key] = resp
 
 
+def _online_request(method, url, data=None,
+                    headers=None, parameters=None):
+    """
+    Handle API errors on conditional requests and try
+    to serve contents from cache
+    """
+    try:
+        resp = requests.request(method=method,
+                                url=url,
+                                headers=headers,
+                                data=data,
+                                timeout=REQUESTS_TIMEOUT,
+                                params=parameters)
+        resp.raise_for_status()
+        return resp, True
+
+    except (requests.exceptions.Timeout, requests.exceptions.HTTPError):
+        return resp, False
+
+
+def _handle_not_changed(cached_response, per_page_elements,
+                        headers, method, url, data, parameters,
+                        cache, cache_key):
+    if len(cached_response.json()) == int(per_page_elements) and \
+       not cached_response.links:
+
+        headers.pop('If-None-Match')
+        resp, valid_response = _online_request(method=method,
+                                               url=url,
+                                               headers=headers,
+                                               data=data,
+                                               parameters=parameters)
+
+        if not valid_response:
+            return _serve_from_cache(resp, cached_response, url)
+
+        LOG.info('ONLINE GET CACHE_MISS %s', url)
+        resp.headers['X-Cache'] = 'ONLINE_MISS'
+        _cache_response(resp, cache, cache_key)
+        return resp
+
+    LOG.info('ONLINE GET CACHE_HIT %s', url)
+    cached_response.headers['X-Cache'] = 'ONLINE_HIT'
+    return cached_response
+
+
+def _serve_from_cache(resp, cached_response, url):
+
+    if cached_response is None:
+        LOG.info('RATE_LIMITED GET CACHE_MISS %s', url)
+        resp.headers['X-Cache'] = 'API_ERROR_CACHE_MISS'
+        return resp
+
+    LOG.info('RATE_LIMITED GET CACHE_HIT %s', url)
```

Comment:
why rate limited? That message should only be logged when we get a 403, isn't it?

Reply:
Forgot to change the log message as I changed the behaviour